Add video modality for InstrucBLIP

zucchini-nlp opened the pull request:
# What does this PR do?

I made these changes a month ago and forgot contributing. This PR adds video processing capabilities for InstructBLIP models. The paper states InstructBLIP was trained and evaluated on video, along with images and the original repo has some code on how video inference works. 

Seems like this feature has some interest from the community (see [here](https://huggingface.co/Salesforce/instructblip-vicuna-7b/discussions/8)), so I believe we can add it.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_30182). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

NielsRogge commented on src/transformers/models/instructblip/image_processing_instructblip.py:
```diff
@@ -0,0 +1,367 @@
+# coding=utf-8
+# Copyright 2022 The HuggingFace Inc. team. All rights reserved.
```
```suggestion
# Copyright 2024 The HuggingFace Inc. team. All rights reserved.
```

NielsRogge commented on src/transformers/models/instructblip/image_processing_instructblip.py:
```diff
@@ -0,0 +1,367 @@
+
+class InstructBlipImageProcessor(BaseImageProcessor):
+    r"""
+    Constructs an Instruct image processor.
```
```suggestion
    Constructs an InstructBLIP image processor.
```

NielsRogge commented on src/transformers/models/instructblip/image_processing_instructblip.py:
```diff
@@ -0,0 +1,367 @@
+
+    model_input_names = ["pixel_values"]
+
+    def __init__(
```
This can perhaps also be Copied from

zucchini-nlp replied:
Okay, let me try. Then I have to add "copy ignore" on the preprocess probably

NielsRogge replied:
No that's only if you add `Copied from` to the class. In that case you can add "Ignore copy" above methods that you don't want to copy

NielsRogge commented on src/transformers/models/instructblip/processing_instructblip.py:
```diff
@@ -57,6 +57,7 @@ def __init__(self, image_processor, tokenizer, qformer_tokenizer):
     def __call__(
         self,
         images: ImageInput = None,
+        videos: ImageInput = None,
```
cc @molbap since we'd like to standardize multimodal processors, this one isn't making it easier 😅 

at some point we will have a `VideoTextToTextPipeline`, and we'll need to make sure they all have the same API.

See also the `ImageTextToTextPipeline` which is worked on at https://github.com/huggingface/transformers/pull/29572

molbap replied:
Sure! Re to normalizing processors, for models taking video inputs, `vivit`, `videomae`, `tvlt` have `videos: ImageInput` but `tvp` has `videos: Union[ImageInput, List[ImageInput], List[List[ImageInput]]],`. However `x_clip` reuses `videomae`'s processor. 

Overall `ImageInput` is defined as a type union of Images or list of images. Looks like in the future we might prefer supporting _at least_ list of list of images so a VideoInput defined as such could make sense, or an union of types as done in x_clip.

NielsRogge reviewed (commented):
Wow that's awesome, thanks for working on that!

I'm just concerned about 2 things:
- making sure that we have a robust API for multimodal processors that is consistent
- the current InstructBLIP models on the hub all use [BlipImageProcessor](https://huggingface.co/Salesforce/instructblip-vicuna-7b/blob/main/preprocessor_config.json#L11). This PR would introduce a new image processor, I guess we would then need to update the [auto mapping](https://github.com/huggingface/transformers/blob/e516d1b19d035469b4852e34ba0356587e6f8ade/src/transformers/models/auto/image_processing_auto.py#L75) to make sure `AutoImageProcessor` still works as expected.

zucchini-nlp commented:
@NielsRogge 

> 1. making sure that we have a robust API for multimodal processors that is consistent
yeah, that needs to be reworked. Right now we have only Blip as first model that supports videos and there willl be VideoLlava. 

Unfortunately video llava processing is going a different way to easily be able to interleave modalities. I guess that is part of what is being discussed internally in slack

> 2. the current InstructBLIP models on the hub all use [BlipImageProcessor](https://huggingface.co/Salesforce/instructblip-vicuna-7b/blob/main/preprocessor_config.json#L11). This PR would introduce a new image processor, I guess we would then need to update the [auto mapping (https://github.com/huggingface/transformers/blob/e516d1b19d035469b4852e34ba0356587e6f8ade/src/transformers/models/auto/image_processing_auto.py#L75) to make sure AutoImageProcessor still works as expected.

Yep, seems like now it works only if calling specifically "InstructBlipImageProcessor"

amyeroberts commented on src/transformers/models/instructblip/image_processing_instructblip.py:
```diff
@@ -0,0 +1,370 @@
+    # Ignore copy
+    def preprocess(
+        self,
+        images: ImageInput = None,
```
If we're adding an image processor for InstructBlip, and it can process images, then it should process images consistently with how they were processed for the previous image processor (blip). 

Whilst the processing steps are the same i.e. the processed image is consistent, the output shape won't be because this will add an extra axis to the output images i.e. they become `(batch_size, num_frames, num_channels, height, width)`. Instead, we should keep the same output shape for images, this will allow us to add this to the `image_processing_auto.py` mapping for instruct blip

amyeroberts reviewed (commented):
Thanks for working on adding this capability! 

Two general comments: 
* We have to be careful here with the mappings wrt backwards compatibility and expected behaviour. As a user, I should be able to do:

```
image_processor = InstructBlipImageProcessor()
images = image_processor(images, return_tensors="pt")
```
and get exactly the same output as I was getting before with the blip image processor

* We should avoid adding lots of if-statements in existing modeling code and instead add a new model

zucchini-nlp commented:
@amyeroberts 
I see, but I am not sure how to make a new model and keep it easy for users to load and for us to maintain. For example: we can make a separate processing for video modality, which will be further enhanced with new features. Call it `InstructBlipVideoProcessing` and another class for video-based generation, `InstructBlipForConditionalVideoGeneration`? For the case of blip, which will never work with interleaving two vision modalities at the same time it might work. But I also want to make sure there will be consistency in how video+image modality LLMs are handled in `transformers`.

What if:
1. Image processor returns one of two possible tensors "pixel_values_image" (4-dim tensor) or "pixel_values_video" (5-dim tensor). This is the way I made it for VideoLlava, which actually can generate from interleaving both at the same time.
    1.1 Or maybe make Image and Video processor as separate classes, and call the appropriate while processing.
2. Modeling stays the same, but we add one line which expands image dimensionality so that it is `(batch_size, 1, num_channels, height, width)` and keep the rest of code as if it were a 1-frame video passed. All the conditions then can be removed, if we treat vision inputs as video-like but 1-frame for image and 4-frame for video clip.

zucchini-nlp commented:
@amyeroberts ping

molbap commented:
@zucchini-nlp one thing related that I'll merge this week, regarding processors: https://github.com/huggingface/transformers/pull/30511 I added a `VideosKwargs`but no `VideosInput` yet. Personal opinion from having spent too much time around `Processors`, I think a separate model would be actually easier to maintain rather than patching a previous one, because _what the model does_ to _which modality_ is easier to understand rather than mixed modalities. I don't have the final say on this, so just a comment :)

amyeroberts commented:
@zucchini-nlp Sorry for the late reply here. 

> Modeling stays the same, but we add one line which expands image dimensionality so that it is (batch_size, 1, num_channels, height, width) and keep the rest of code as if it were a 1-frame video passed. All the conditions then can be removed, if we treat vision inputs as video-like but 1-frame for image and 4-frame for video clip.

This is a neat solution. My main question would be, what does this do to the shapes of the returned tensors e.g. `hidden_states`? If they remain the same, then this is a nice easy way to enable this within the modeling file. 

If they're not the same, then we'll need to add a new class in the modeling file e.g. `InstructBlipForVideoConditionalGeneration` or perhaps, have a new modeling file `src/transformers/models/instructblip/modeling_instructblip_video.py`. The latter would add a new model type e.f. `VideoInstructBlip` / `InstructBlipVideo` which would enable the correct auto mapping in `image_processing_auto.py`. It's not unheard of for us to add new models which can load existing checkpoints under a new architecture name. 

In terms of the image processor: 
* If we can use the same class and just expand within the forward pass, then the trick is to correctly batch the inputs such that for a batch of images, the shape of the output pixel values remains `(batch_size, num_channels, height, width)` and for videos it's `(batch_size, num_frames, num_channels, height, width)`. Both should be output as `pixel_values`. That is, the output of the image processor for images should remain unchanged from the previous behaviour: same shape, and same key names in the `BatchFeature` output
* If we can't use the same class, but add a single module in `modeling_instructblip` you can do as above
* If we can't use the same class, but add a new modeling file, then we can add a new separate image processor. 

My vote would be for a new model. Having to handle video and images within the same image processor is a good indication they should be separated

zucchini-nlp commented:
Thanks for detailed explanations! At this point it should be possible to go with the first option, I will have to go back and check. If not, making a separate model sounds good since BLIP will never work with both modalities at the same time.

@molbap i see, probably having separate VideoProcessor files can be a better solutions for mutli-modal models, instead of packing it all in ImageProcessor

zucchini-nlp commented:
I went for the "let's keep one processing and modeling files" way. Currently the following changes are applied:

1. Processor can accept as arg either `images` or `videos`
2. Image processor returns pixel values of shape (b, c, h, w) for images by squeezing the extra dim and adds and extra frame dimension for videos.
3. Modeling file unsqueezed back the frame dimension for images, and continues running as if the inputs are all videos. Finally the frame dimension is merged back to embeddings' sequence length.

Slow tests are passing locally + added a few tests for the video generation.

amyeroberts commented:
Thanks for the continued work on this and apologies for the delay in my review. Skimming over the PR, and thinking about this more I think we should split this up such that we have a video image processor and a video model. We want to avoid conditional outputs from our base processing objects as well as conditional logic within our model's forward passes. As we won't interleave images and videos i.e. we don't expect a user to both be using video and images at the same time then we don't need to be able to handle these all at once

zucchini-nlp committed:
squash in single commit

zucchini-nlp committed:
add docs

zucchini-nlp committed:
dummy obj

zucchini-nlp committed:
more changes in diff converter

zucchini-nlp committed:
Merge branch 'main' into blip_video

zucchini-nlp committed:
tiny fix

zucchini-nlp committed:
make docs happy

zucchini-nlp commented:
@amyeroberts oke, I made Video InstructBlip its own model with its separate modeling and image processing files. Added some tests and indicated in the model-doc that it's the same model as InstructBlip except for the video processing capability. Ready for review

@ArthurZucker I made more changes into diff converter in this PR, as it didn't work in some cases. Specifically:

- Some models apply CamelCase without an underscore that splits subwords, like InstructBlip, Diff converter cannot infer correct model name in this case, so I added a possibility to indicate model-names by passing it as an arg for converter
- In case if we have a config that inherits from another config in the library, the auto-generated config doesn't get all the imports from its parent class causing errors. I added a special visited_module_config to store config-specific classes. I could use the same `visited_module` that is already defined but it will result in configs being imported into the config file
- Sometimes configs need globally defined vars, e.g. in case of the logger. So there's a new line that ports "SimpleStatements" to configs. If they are not used, everything will be cleaned up by ruff anyway 
- Properties with their setters weren't being ported because they have identical naming, so the fix is to iterate "node.body" directly w/o saving it in a dict

Plus these are the changes in [LLaVa-NeXT-Video PR](https://github.com/huggingface/transformers/pull/31252), just duplicating to have all written down somewhere :)
- Sometimes we want to add new methods in a class, and still retain all methods from the parent. This case wasn't covered, I added a few lines to fix it
- Inferring model-name from file-name didn't consider long model names with an underscore, fixed by modifying the regex

Still have to fix cases when only "init" docstring is changed, I couldn't make it work yet

zucchini-nlp committed:
skip test

zucchini-nlp committed:
repo consistency tests

amyeroberts reviewed (commented):
Looks great - thanks for splitting this up and adding this model! 

Just a few comment and some questions about the changes to the diff_converter. Main comment is about making sure generate is properly tested

zucchini-nlp commented on tests/models/instructblipvideo/test_modeling_instructblipvideo.py:
```diff
@@ -0,0 +1,595 @@
+class InstructBlipVideoForConditionalGenerationDecoderOnlyTest(
+    ModelTesterMixin, GenerationTesterMixin, unittest.TestCase
+):
+    all_model_classes = (InstructBlipVideoForConditionalGeneration,) if is_torch_available() else ()
```
GenerationMixin still can't work for VLMs, and I am planning to properly add it after some unification of VLM processors. Otherwise we'll have so many conditional checks inside testing

zucchini-nlp commented on utils/diff_model_converter.py:
```diff
@@ -356,16 +358,19 @@ def replace_call_to_super(class_finder: ClassFinder, updated_node: cst.ClassDef,
         self.transformers_imports = {}      # maps the imports name like "from transformers.models.xxx" to the parsed AST module
         self.imported_mapping = {}          # stores the name of the imported classes, with their source {"LlamaModel":"transformers.model.llama.modeling_llama"}
         self.visited_module = {}            # modules visited like "transformers.models.llama.modeling_llama"
+        self.visited_module_config = {}     # modules visited like "transformers.models.llama.modeling_llama" in config file, needed to not mix config vs modeling imports
```
Sorry, maybe needs another example. This was needed for me because the auto-generated config wasn't getting imports from its "super config". In other words, it was only copying imports from diff files and removing unused ones by ruff.

One solution may be to indicate all imports in diff, but if they aren't used ruff removes them eventually. In my case `PreTrainedConfig` wasn't being imported for ex

zucchini-nlp commented on utils/diff_model_converter.py:
```diff
@@ -551,10 +565,20 @@ def convert_file(diff_file, cst_transformers=None):
         nargs="+",
         help="A list of `diff_xxxx` files that should be converted to single model file",
     )
+    parser.add_argument(
+        "--old_model_name",
+        required=False,
+        help="The name of the model from which the copying is done in CamelCase. If not provided is inferred from diff-file",
+    )
+    parser.add_argument(
+        "--new_model_name",
+        required=False,
+        help="The name of the new model being added in CamelCase. If not provided is inferred from diff-file",
+    )
```
This is for models that do camel case without underscores, like InstructBLIP (instructblip) vs LlavaNext (llava_next). In second case we can infer where to make a capital letter, while in former it's impossible so I decided to give users freedom passing model names

zucchini-nlp commented on utils/diff_model_converter.py:
```diff
@@ -474,7 +485,7 @@ def leave_ClassDef(self, original_node, updated_node):
             for dependency, _ in list_dependencies:
                 node = class_finder.global_nodes.get(dependency, None)
-                if node is not None:
+                if node is not None and "Config" not in class_name:
```
This is added because diff importing configs to the configuration files, even though they are defined as a class a few lines below

zucchini-nlp committed:
update docstring

zucchini-nlp committed:
style

zucchini-nlp committed:
Merge remote-tracking branch 'upstream/main' into blip_video

zucchini-nlp committed:
fix tests

ArthurZucker reviewed (commented):
diff converter changes look nice IMO, but we should not need to import all the classes. New diff converter is able to parse dependencies so tell me if this is not the case!

zucchini-nlp committed:
change diff imports

zucchini-nlp committed:
Merge branch 'huggingface:main' into blip_video

zucchini-nlp commented:
@amyeroberts this one is ready for the final review I guess

amyeroberts reviewed (approved):
Another great addition - thanks! 

As with Llava Next Video, we just need a final run of the slow tests and we're good to merge ❤️

zucchini-nlp committed:
[run-slow] instructblipvideo

zucchini-nlp committed:
[run-slow] instructblipvideo

zucchini-nlp committed:
fix tests and remove logit check

zucchini-nlp committed:
[run-slow] instructblipvideo

zucchini-nlp commented:
Got the CI green, including slow tests. Will merge the PR
